File: src/main/java/raccoonlang/parser/Parser.java
package raccoonlang.parser;

import raccoonlang.parser.ast.FileAstNode;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenStream;
import raccoonlang.tokenizer.TokenType;

public class Parser {

    public class ParserState {
        private int tokenIndex = 0;

        public ParserState(int tokenIndex){
            this.tokenIndex = tokenIndex;
        }
    }

    private TokenStream tokenStream;

    public Parser(TokenStream tokenStream) {
        this.tokenStream = tokenStream;
    }

    public Token peek() {
        return peek(0);
    }

    public Token peek(int lookAhead) {
        return tokenStream.peek(lookAhead);
    }

    public Token take(TokenType tokenType) {
        return tokenStream.take(tokenType);
    }

    public Token take() {
        return tokenStream.take();
    }

    public ParserState shelfState(){
        return new ParserState(tokenStream.getPosition());
    }

    public void restore(ParserState parserState){
        tokenStream.seek(parserState.tokenIndex);
    }

    public static FileAstNode parse(TokenStream tokenStream) {
        Parser parser = new Parser(tokenStream);
        FileAstNode fileAstNode = FileAstNode.parse(parser);
        parser.take(TokenType.EOF);
        return fileAstNode;
    }
}

File: Parsing/Parser.cs
namespace Raccoonlang.Parsing;

using Tokenizing;
using AST;

public class Parser
{
    public class ParserState
    {
        public int TokenIndex {get;}

        public ParserState(int tokenIndex)
        {
            this.TokenIndex = tokenIndex;
        }
    }

    private TokenStream _tokenStream;

    public Parser(TokenStream stream)
    {
        this._tokenStream = stream;
    }

    public void Skip()
    {
        this._tokenStream.Skip();
    }

    public void Skip(int skipAmount)
    {
        this._tokenStream.Skip(skipAmount);
    }

        public Token Peek()
    {
        return Peek(0);
    }

    public Token Peek(int peek) 
    {
        return this._tokenStream.Peek(peek);
    }

    public Token Take()
    {
        return this._tokenStream.Take();
    }

    public Token Take(TokenType tokenType) 
    {
        return this._tokenStream.Take(tokenType);
    }

    public ParserState ShelfState()
    {
        return new ParserState(_tokenStream.Position);
    }

    public void RestoreState(ParserState state)
    {
        _tokenStream.Seek(state.TokenIndex);
    }

    public void PrintCurrent()
    {
        Console.WriteLine(this._tokenStream.Current());
    }

    public static FileAstNode Parse(TokenStream stream)
    {
        Parser parser = new Parser(stream);
        FileAstNode node = FileAstNode.Parse(parser);
        //parser.Take(TokenType.EOF);
        return node;
    }
}